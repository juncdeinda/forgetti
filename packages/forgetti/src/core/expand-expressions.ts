import * as t from '@babel/types';
import type * as babel from '@babel/core';
import type { ComponentNode, StateContext } from './types';
import { getHookCallType } from './get-hook-call-type';
import { isPathValid } from './checks';
import unwrapNode from './unwrap-node';

function isStatementValid(path: babel.NodePath): boolean {
  if (path) {
    switch (path.node.type) {
      case 'ForInStatement':
      case 'ForOfStatement':
      case 'ForStatement':
      case 'WhileStatement':
      case 'DoWhileStatement':
        return false;
      default:
        return true;
    }
  }
  return false;
}

function isInValidExpression(path: babel.NodePath): boolean {
  let current = path.parentPath;
  let prev = path;
  while (current) {
    if (
      isPathValid(current, t.isConditionalExpression) &&
      (current.get('consequent').node === prev.node ||
        current.get('alternate').node === prev.node)
    ) {
      return false;
    }
    if (
      isPathValid(current, t.isLogicalExpression) &&
      current.get('right').node === prev.node
    ) {
      return false;
    }
    prev = current;
    current = current.parentPath;
  }
  return true;
}

const UNDEFINED_LITERAL = t.unaryExpression('void', t.numericLiteral(0));

function transformOptionalCall(
  path: babel.NodePath<t.OptionalCallExpression>,
): t.Expression {
  const unwrappedID = unwrapNode(path.node.callee, t.isIdentifier);
  if (unwrappedID) {
    return t.conditionalExpression(
      t.binaryExpression('==', unwrappedID, t.nullLiteral()),
      UNDEFINED_LITERAL,
      t.callExpression(unwrappedID, path.node.arguments),
    );
  }
  const temp = path.scope.generateUidIdentifier('nullish');
  path.scope.push({
    kind: 'let',
    id: temp,
  });
  const unwrappedCallee =
    unwrapNode(path.node.callee, t.isMemberExpression) ||
    unwrapNode(path.node.callee, t.isOptionalMemberExpression);
  if (unwrappedCallee) {
    let unwrappedObject = unwrapNode(unwrappedCallee.object, t.isIdentifier);
    if (!unwrappedObject) {
      unwrappedObject = path.scope.generateUidIdentifier('object');
      path.scope.push({
        kind: 'let',
        id: unwrappedObject,
      });
      unwrappedCallee.object = t.assignmentExpression(
        '=',
        unwrappedObject,
        unwrappedCallee.object,
      );
    }
    return t.conditionalExpression(
      t.binaryExpression(
        '==',
        t.assignmentExpression('=', temp, unwrappedCallee),
        t.nullLiteral(),
      ),
      UNDEFINED_LITERAL,
      t.callExpression(t.memberExpression(temp, t.identifier('call')), [
        unwrappedObject,
        ...path.node.arguments,
      ]),
    );
  }
  return t.conditionalExpression(
    t.binaryExpression(
      '==',
      t.assignmentExpression('=', temp, path.node.callee),
      t.nullLiteral(),
    ),
    UNDEFINED_LITERAL,
    t.callExpression(temp, path.node.arguments),
  );
}

function transformOptionalMember(
  path: babel.NodePath<t.OptionalMemberExpression>,
): t.Expression {
  const unwrappedID = unwrapNode(path.node.object, t.isIdentifier);
  if (unwrappedID) {
    return t.conditionalExpression(
      t.binaryExpression('==', unwrappedID, t.nullLiteral()),
      UNDEFINED_LITERAL,
      t.memberExpression(unwrappedID, path.node.property, path.node.computed),
    );
  }
  const temp = path.scope.generateUidIdentifier('nullish');
  path.scope.push({
    kind: 'let',
    id: temp,
  });
  return t.conditionalExpression(
    t.binaryExpression(
      '==',
      t.assignmentExpression('=', temp, path.node.object),
      t.nullLiteral(),
    ),
    UNDEFINED_LITERAL,
    t.memberExpression(temp, path.node.property, path.node.computed),
  );
}

export function expandExpressions(
  ctx: StateContext,
  path: babel.NodePath<ComponentNode>,
): void {
  if (
    path.node.type === 'ArrowFunctionExpression' &&
    path.node.body.type !== 'BlockStatement'
  ) {
    path.node.body = t.blockStatement([t.returnStatement(path.node.body)]);
  }
  path.traverse({
    OptionalCallExpression(p) {
      const parent = p.getFunctionParent();
      const statement = p.getStatementParent();

      if (parent === path && statement) {
        p.replaceWith(transformOptionalCall(p));
      }
    },
    OptionalMemberExpression(p) {
      const parent = p.getFunctionParent();
      const statement = p.getStatementParent();

      if (parent === path && statement) {
        p.replaceWith(transformOptionalMember(p));
      }
    },
    AssignmentExpression(p) {
      const parent = p.getFunctionParent();
      const statement = p.getStatementParent();

      if (
        parent === path &&
        statement &&
        isStatementValid(statement) &&
        isInValidExpression(p) &&
        !isPathValid(p.parentPath, t.isStatement) &&
        !isPathValid(p.parentPath, t.isVariableDeclarator)
      ) {
        const id = p.scope.generateUidIdentifier('hoisted');
        statement.insertBefore(
          t.variableDeclaration('let', [t.variableDeclarator(id, p.node)]),
        );
        p.replaceWith(id);
      }
    },
    CallExpression(p) {
      const parent = p.getFunctionParent();
      const statement = p.getStatementParent();

      if (
        parent === path &&
        statement &&
        !isPathValid(p.parentPath, t.isStatement) &&
        !isPathValid(p.parentPath, t.isVariableDeclarator)
      ) {
        const hookType = getHookCallType(ctx, p);
        if (hookType === 'custom' || hookType === 'effect') {
          const id = p.scope.generateUidIdentifier('hoisted');
          statement.insertBefore(
            t.variableDeclaration('let', [t.variableDeclarator(id, p.node)]),
          );
          p.replaceWith(id);
        }
      }
    },
  });

  path.scope.crawl();
}
