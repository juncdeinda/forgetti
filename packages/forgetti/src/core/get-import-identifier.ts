import type * as babel from '@babel/core';
import type * as t from '@babel/types';
import { addDefault, addNamed } from '@babel/helper-module-imports';
import type { ImportDefinition } from './presets';
import type { StateContext } from './types';

export default function getImportIdentifier(
  ctx: StateContext,
  path: babel.NodePath,
  definition: ImportDefinition,
): t.Identifier {
  const name = definition.kind === 'default' ? 'default' : definition.name;
  const target = `${definition.source}[${name}]`;
  const current = ctx.imports.get(target);
  if (current) {
    return current;
  }
  const newID =
    definition.kind === 'named'
      ? addNamed(path, definition.name, definition.source)
      : addDefault(path, definition.source);
  ctx.imports.set(target, newID);
  return newID;
}
