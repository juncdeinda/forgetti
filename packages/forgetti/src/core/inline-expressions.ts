import type * as babel from '@babel/core';
import * as t from '@babel/types';
import type { ComponentNode } from './types';
import { isPathValid } from './checks';

function isInValidExpression(path: babel.NodePath): boolean {
  let current = path.parentPath;
  let prev = path;
  while (current) {
    if (
      isPathValid(current, t.isConditionalExpression) &&
      (current.get('consequent').node === prev.node ||
        current.get('alternate').node === prev.node)
    ) {
      return false;
    }
    if (
      isPathValid(current, t.isLogicalExpression) &&
      current.get('right').node === prev.node
    ) {
      return false;
    }
    prev = current;
    current = current.parentPath;
  }
  return true;
}

function inlineExpression(
  parentPath: babel.NodePath<ComponentNode>,
  path: babel.NodePath<t.Expression>,
) {
  if (
    path.getFunctionParent() === parentPath &&
    isPathValid(path, t.isIdentifier)
  ) {
    const binding = path.scope.getBinding(path.node.name);
    if (binding?.referenced && binding.referencePaths.length === 1) {
      switch (binding.kind) {
        case 'const':
        case 'let':
        case 'var': {
          // move the node to the reference
          const ref = binding.referencePaths[0];
          if (
            isInValidExpression(ref) &&
            isPathValid(binding.path, t.isVariableDeclarator) &&
            binding.path.node.init &&
            isPathValid(binding.path.get('id'), t.isIdentifier) &&
            binding.path.scope.getBlockParent() === ref.scope.getBlockParent()
          ) {
            ref.replaceWith(binding.path.node.init);
            binding.path.remove();
          }
          break;
        }
      }
    }
  }
}

export function inlineExpressions(path: babel.NodePath<ComponentNode>): void {
  path.traverse({
    Expression(p) {
      inlineExpression(path, p);
    },
  });
  path.scope.crawl();
}
