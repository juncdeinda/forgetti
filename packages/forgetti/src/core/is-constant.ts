import type * as babel from '@babel/core';
import * as t from '@babel/types';
import { isNestedExpression, shouldSkipNode, isPathValid } from './checks';
import type OptimizerScope from './optimizer-scope';
import getForeignBindings, { isForeignBinding } from './get-foreign-bindings';
import type { ComponentNode, StateContext } from './types';
import { getHookCallType } from './get-hook-call-type';

interface OptimizerInstance {
  ctx: StateContext;
  scope: OptimizerScope;
  path: babel.NodePath<ComponentNode>;
  nodeConstantCheckCache: WeakMap<babel.NodePath<t.Expression>, boolean>;
}

function isTemplateLiteralConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.TemplateLiteral>,
): boolean {
  const expressions = path.get('expressions');
  for (let i = 0, len = expressions.length; i < len; i++) {
    const expr = expressions[i];
    if (isPathValid(expr, t.isExpression)) {
      if (!isConstant(instance, expr)) {
        return false;
      }
    } else {
      return false;
    }
  }
  return true;
}

function isIdentifierConstant(
  instance: OptimizerInstance,
  path: babel.NodePath,
  node: t.Identifier,
): boolean {
  switch (node.name) {
    case 'undefined':
    case 'NaN':
    case 'Infinity':
      return true;
    default: {
      if (isForeignBinding(instance.path, path, node.name)) {
        return true;
      }
      const binding = path.scope.getBindingIdentifier(node.name);
      if (binding) {
        return instance.scope.hasConstant(binding);
      }
      return true;
    }
  }
}

function isMemberExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.MemberExpression>,
): boolean {
  if (isConstant(instance, path.get('object'))) {
    if (path.node.computed) {
      const property = path.get('property');
      if (isPathValid(property, t.isExpression)) {
        return isConstant(instance, property);
      }
    }
    return true;
  }
  return false;
}

function isConditionalExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.ConditionalExpression>,
): boolean {
  return (
    isConstant(instance, path.get('test')) &&
    isConstant(instance, path.get('consequent')) &&
    isConstant(instance, path.get('alternate'))
  );
}

function isBinaryExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.BinaryExpression>,
): boolean {
  const left = path.get('left');
  const right = path.get('right');
  return (
    isPathValid(left, t.isExpression) &&
    isConstant(instance, left) &&
    isConstant(instance, right)
  );
}

function isLogicalExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.LogicalExpression>,
): boolean {
  return (
    isConstant(instance, path.get('left')) &&
    isConstant(instance, path.get('right'))
  );
}

function isUnaryExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.UnaryExpression>,
): boolean {
  switch (path.node.operator) {
    case 'throw':
    case 'delete':
      return false;
    default:
      return isConstant(instance, path.get('argument'));
  }
}

function isCallExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.CallExpression>,
): boolean {
  const hookType = getHookCallType(instance.ctx, path);
  if (hookType === 'none') {
    const callee = path.get('callee');
    if (isPathValid(callee, t.isExpression) && !isConstant(instance, callee)) {
      return false;
    }
    const args = path.get('arguments');
    for (let i = 0, len = args.length; i < len; i++) {
      const arg = args[i];
      if (isPathValid(arg, t.isExpression) && !isConstant(instance, arg)) {
        return false;
      }
      if (
        isPathValid(arg, t.isSpreadElement) &&
        !isConstant(instance, arg.get('argument'))
      ) {
        return false;
      }
    }
    return true;
  }
  return false;
}

function isFunctionExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.FunctionExpression | t.ArrowFunctionExpression>,
): boolean {
  const bindings = getForeignBindings(path);
  for (let i = 0, len = bindings.length; i < len; i++) {
    const binding = bindings[i];
    if (!isIdentifierConstant(instance, path, binding)) {
      return false;
    }
  }
  return true;
}

function isArrayExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.ArrayExpression | t.TupleExpression>,
): boolean {
  const elements = path.get('elements');
  for (let i = 0, len = elements.length; i < len; i++) {
    const element = elements[i];
    if (isPathValid(element, t.isSpreadElement)) {
      if (!isConstant(instance, element.get('argument'))) {
        return false;
      }
    }
    if (isPathValid(element, t.isExpression)) {
      if (!isConstant(instance, element)) {
        return false;
      }
    }
  }
  return true;
}

function isObjectPropertyConstant(
  instance: OptimizerInstance,
  property: babel.NodePath<t.ObjectProperty>,
): boolean {
  if (property.node.computed) {
    const key = property.get('key');
    if (isPathValid(key, t.isExpression) && !isConstant(instance, key)) {
      return false;
    }
  }
  const value = property.get('value');
  if (isPathValid(value, t.isExpression) && !isConstant(instance, value)) {
    return false;
  }
  return true;
}

function isObjectMethodConstant(
  instance: OptimizerInstance,
  property: babel.NodePath<t.ObjectMethod>,
): boolean {
  const bindings = getForeignBindings(property);
  for (let k = 0, klen = bindings.length; k < klen; k++) {
    const binding = bindings[k];
    if (!isIdentifierConstant(instance, property, binding)) {
      return false;
    }
  }
  return true;
}

function isObjectExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.ObjectExpression | t.RecordExpression>,
): boolean {
  const properties = path.get('properties');
  for (let i = 0, len = properties.length; i < len; i++) {
    const property = properties[i];
    if (
      isPathValid(property, t.isSpreadElement) &&
      !isConstant(instance, property.get('argument'))
    ) {
      return false;
    }
    if (
      isPathValid(property, t.isObjectProperty) &&
      !isObjectPropertyConstant(instance, property)
    ) {
      return false;
    }
    if (
      isPathValid(property, t.isObjectMethod) &&
      !isObjectMethodConstant(instance, property)
    ) {
      return false;
    }
  }
  return true;
}

function isNewExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.NewExpression>,
): boolean {
  const callee = path.get('callee');
  if (isPathValid(callee, t.isExpression) && isConstant(instance, callee)) {
    const args = path.get('arguments');
    for (let i = 0, len = args.length; i < len; i++) {
      const arg = args[i];
      if (isPathValid(arg, t.isExpression) && !isConstant(instance, arg)) {
        return false;
      }
      if (
        isPathValid(arg, t.isSpreadElement) &&
        !isConstant(instance, arg.get('argument'))
      ) {
        return false;
      }
    }
    return true;
  }
  return false;
}

function isSequenceExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.SequenceExpression>,
): boolean {
  const exprs = path.get('expressions');
  for (let i = 0, len = exprs.length; i < len; i++) {
    if (!isConstant(instance, exprs[i])) {
      return false;
    }
  }
  return true;
}

function isTaggedTemplateExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.TaggedTemplateExpression>,
): boolean {
  return (
    isConstant(instance, path.get('tag')) &&
    isConstant(instance, path.get('quasi'))
  );
}

function isJSXExpressionContainerConstant(
  instance: OptimizerInstance,
  child: babel.NodePath<t.JSXExpressionContainer>,
): boolean {
  const expr = child.get('expression');
  return isPathValid(expr, t.isExpression) && isConstant(instance, expr);
}

function isJSXChildrenConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.JSXFragment | t.JSXElement>,
): boolean {
  const children = path.get('children');
  for (let i = 0, len = children.length; i < len; i++) {
    const child = children[i];
    if (
      isPathValid(child, t.isJSXElement) &&
      !isJSXElementConstant(instance, child)
    ) {
      return false;
    }
    if (
      isPathValid(child, t.isJSXFragment) &&
      !isJSXChildrenConstant(instance, child)
    ) {
      return false;
    }
    if (
      isPathValid(child, t.isJSXSpreadChild) &&
      !isConstant(instance, child.get('expression'))
    ) {
      return false;
    }
    if (
      isPathValid(child, t.isJSXExpressionContainer) &&
      !isJSXExpressionContainerConstant(instance, child)
    ) {
      return false;
    }
  }
  return true;
}

function isJSXNameConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<
    t.JSXIdentifier | t.JSXMemberExpression | t.JSXNamespacedName
  >,
): boolean {
  if (isPathValid(path, t.isJSXNamespacedName)) {
    return true;
  }
  if (isPathValid(path, t.isJSXMemberExpression)) {
    return isJSXNameConstant(instance, path.get('object'));
  }
  if (isPathValid(path, t.isJSXIdentifier)) {
    if (/^[A-Z]/.test(path.node.name)) {
      return isForeignBinding(instance.path, path, path.node.name);
    }
  }
  return true;
}

function isJSXAttributeConstant(
  instance: OptimizerInstance,
  attr: babel.NodePath<t.JSXAttribute>,
): boolean {
  const value = attr.get('value');
  if (
    isPathValid(value, t.isJSXElement) &&
    !isJSXElementConstant(instance, value)
  ) {
    return false;
  }
  if (
    isPathValid(value, t.isJSXFragment) &&
    !isJSXChildrenConstant(instance, value)
  ) {
    return false;
  }
  if (isPathValid(value, t.isJSXExpressionContainer)) {
    const expr = value.get('expression');
    if (isPathValid(expr, t.isExpression) && !isConstant(instance, expr)) {
      return false;
    }
  }
  return true;
}

function isJSXOpeningElementConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.JSXOpeningElement>,
): boolean {
  if (isJSXNameConstant(instance, path.get('name'))) {
    const attrs = path.get('attributes');
    for (let i = 0, len = attrs.length; i < len; i++) {
      const attr = attrs[i];
      if (
        isPathValid(attr, t.isJSXAttribute) &&
        !isJSXAttributeConstant(instance, attr)
      ) {
        return false;
      }
      if (
        isPathValid(attr, t.isJSXSpreadAttribute) &&
        !isConstant(instance, attr.get('argument'))
      ) {
        return false;
      }
    }
    return true;
  }
  return false;
}

function isJSXElementConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.JSXElement>,
): boolean {
  return (
    isJSXOpeningElementConstant(instance, path.get('openingElement')) &&
    isJSXChildrenConstant(instance, path)
  );
}

function uncachedIsConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.Expression>,
): boolean {
  if (shouldSkipNode(path.node)) {
    return false;
  }
  if (isPathValid(path, isNestedExpression)) {
    return isConstant(instance, path.get('expression'));
  }
  if (isPathValid(path, t.isLiteral)) {
    if (isPathValid(path, t.isTemplateLiteral)) {
      return isTemplateLiteralConstant(instance, path);
    }
    return true;
  }
  if (isPathValid(path, t.isIdentifier)) {
    return isIdentifierConstant(instance, path, path.node);
  }
  if (isPathValid(path, t.isMemberExpression)) {
    return isMemberExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isConditionalExpression)) {
    return isConditionalExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isBinaryExpression)) {
    return isBinaryExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isLogicalExpression)) {
    return isLogicalExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isUnaryExpression)) {
    return isUnaryExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isCallExpression)) {
    return isCallExpressionConstant(instance, path);
  }
  if (
    isPathValid(path, t.isFunctionExpression) ||
    isPathValid(path, t.isArrowFunctionExpression)
  ) {
    return isFunctionExpressionConstant(instance, path);
  }
  if (
    isPathValid(path, t.isArrayExpression) ||
    isPathValid(path, t.isTupleExpression)
  ) {
    return isArrayExpressionConstant(instance, path);
  }
  if (
    isPathValid(path, t.isObjectExpression) ||
    isPathValid(path, t.isRecordExpression)
  ) {
    return isObjectExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isNewExpression)) {
    return isNewExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isSequenceExpression)) {
    return isSequenceExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isTaggedTemplateExpression)) {
    return isTaggedTemplateExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isJSXFragment)) {
    return isJSXChildrenConstant(instance, path);
  }
  if (isPathValid(path, t.isJSXElement)) {
    return isJSXElementConstant(instance, path);
  }
  if (isPathValid(path, t.isImport)) {
    return true;
  }
  if (isPathValid(path, t.isMetaProperty)) {
    return true;
  }
  return false;
}

export default function isConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.Expression>,
): boolean {
  if (instance.nodeConstantCheckCache.has(path)) {
    return instance.nodeConstantCheckCache.get(path)!;
  }
  const result = uncachedIsConstant(instance, path);
  instance.nodeConstantCheckCache.set(path, result);
  return result;
}
