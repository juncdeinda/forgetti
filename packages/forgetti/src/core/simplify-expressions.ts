import type * as babel from '@babel/core';
import * as t from '@babel/types';
import type { ComponentNode } from './types';

type LiteralState = 'truthy' | 'falsy' | 'nullish' | 'indeterminate';

function getBooleanishState(node: t.Expression): LiteralState {
  switch (node.type) {
    case 'BooleanLiteral':
      return node.value ? 'truthy' : 'falsy';
    case 'NumericLiteral':
      return node.value === 0 ? 'falsy' : 'truthy';
    case 'NullLiteral':
      return 'nullish';
    case 'StringLiteral':
      return node.value === '' ? 'falsy' : 'truthy';
    case 'BigIntLiteral':
      return node.value === '0' ? 'falsy' : 'truthy';
    case 'Identifier': {
      return 'indeterminate';
    }
    // case 'ArrayExpression':
    // case 'FunctionExpression':
    // case 'ArrowFunctionExpression':
    // case 'RegExpLiteral':
    // case 'ObjectExpression':
    // case 'RecordExpression':
    // case 'TupleExpression':
    //   return 'truthy';
    case 'ParenthesizedExpression':
    case 'TypeCastExpression':
    case 'TSAsExpression':
    case 'TSSatisfiesExpression':
    case 'TSNonNullExpression':
    case 'TSTypeAssertion':
    case 'TSInstantiationExpression':
      return getBooleanishState(node.expression);
    default:
      return 'indeterminate';
  }
}

export function simplifyExpressions(path: babel.NodePath<ComponentNode>): void {
  path.traverse({
    ConditionalExpression: {
      exit(p) {
        const state = getBooleanishState(p.node.test);
        if (state === 'truthy') {
          p.replaceWith(p.node.consequent);
        } else if (state !== 'indeterminate') {
          p.replaceWith(p.node.alternate);
        }
      },
    },
    LogicalExpression: {
      exit(p) {
        switch (getBooleanishState(p.node.left)) {
          case 'nullish': {
            p.replaceWith(
              p.node.operator === '??' ? p.node.right : p.node.left,
            );
            break;
          }
          case 'falsy': {
            p.replaceWith(
              p.node.operator === '||' ? p.node.right : p.node.left,
            );
            break;
          }
          case 'truthy': {
            p.replaceWith(
              p.node.operator === '&&' ? p.node.right : p.node.left,
            );
            break;
          }
        }
      },
    },
    UnaryExpression: {
      exit(p) {
        const state = getBooleanishState(p.node.argument);
        switch (p.node.operator) {
          case 'void': {
            if (state !== 'indeterminate') {
              p.replaceWith(t.unaryExpression('void', t.numericLiteral(0)));
            }
            break;
          }
          case '!': {
            if (state === 'truthy') {
              p.replaceWith(t.booleanLiteral(false));
            } else if (state !== 'indeterminate') {
              p.replaceWith(t.booleanLiteral(true));
            }
            break;
          }
        }
      },
    },
    IfStatement: {
      exit(p) {
        const state = getBooleanishState(p.node.test);
        if (state === 'truthy') {
          p.replaceWith(p.node.consequent);
        } else if (state === 'indeterminate') {
          // TODO Should simplify IfStatement?
        } else if (p.node.alternate) {
          p.replaceWith(p.node.alternate);
        } else {
          p.remove();
        }
      },
    },
    WhileStatement: {
      exit(p) {
        if (getBooleanishState(p.node.test) === 'falsy') {
          p.remove();
        }
      },
    },
  });
}
