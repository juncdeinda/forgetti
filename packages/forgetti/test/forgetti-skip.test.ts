import * as babel from '@babel/core';
import { describe, it } from 'vitest';
import type { Options } from '../src';
import plugin from '../src';

const options: Options = {
  preset: 'react',
};

async function compile(code: string): Promise<string> {
  const result = await babel.transformAsync(code, {
    plugins: [[plugin, options]],
    parserOpts: {
      plugins: ['jsx'],
    },
  });

  return result?.code ?? '';
}

describe('forgetti skip', () => {
  it('should optimize non-skipped function declaration', async ({ expect }) => {
    const code = `
function Example(props) {
  return <h1 className={props.className}>{props.children}</h1>;
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });

  it('should skip skipped function declaration', async ({ expect }) => {
    const code = `
/* @forgetti skip */
function Example(props) {
  return <h1 className={props.className}>{props.children}</h1>;
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });

  it('should optimize non-skipped function expression', async ({ expect }) => {
    const code = `
const Example = function (props) {
  return <h1 className={props.className}>{props.children}</h1>;
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });

  it('should skip skipped function expression', async ({ expect }) => {
    const code = `
const /* @forgetti skip */ ExampleA = function (props) {
  return <h1 className={props.className}>{props.children}</h1>;
}
/* @forgetti skip */ const ExampleB = function (props) {
  return <h1 className={props.className}>{props.children}</h1>;
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });

  it('should optimize non-skipped variable declaration', async ({ expect }) => {
    const code = `
const Example = props => {
  return <h1 className={props.className}>{props.children}</h1>;
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });

  it('should skip skipped variable declaration', async ({ expect }) => {
    const code = `
const /* @forgetti skip */ ExampleA = props => {
  return <h1 className={props.className}>{props.children}</h1>;
}
/* @forgetti skip */ const ExampleB = props => {
  return <h1 className={props.className}>{props.children}</h1>;
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
});
