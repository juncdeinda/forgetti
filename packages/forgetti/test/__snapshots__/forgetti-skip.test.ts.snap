// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`forgetti skip > should optimize non-skipped function declaration 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
import { memo as _memo } from \\"react\\";
import { $$memo as _$$memo } from \\"forgetti/runtime\\";
const _Example = _$$memo(_memo, \\"_Example\\", _values => <h1 className={_values[0]}>{_values[1]}</h1>);
function Example(props) {
  let _cache = _$$cache(_useMemo, 8),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = _value.className,
    _equals2 = _$$equals(_cache, 2, _value2),
    _value3 = _equals2 ? _cache[2] : _cache[2] = _value2,
    _value4 = _equals ? _cache[3] : _cache[3] = _value.children,
    _equals3 = _$$equals(_cache, 4, _value4),
    _value5 = _equals3 ? _cache[4] : _cache[4] = _value4,
    _value6 = _equals2 && _equals3 ? _cache[5] : _cache[5] = [_value3, _value5],
    _equals5 = _$$equals(_cache, 6, _value6),
    _value7 = _equals5 ? _cache[6] : _cache[6] = _value6;
  return /*@forgetti jsx*/_equals5 ? _cache[7] : _cache[7] = <_Example v={_value7} />;
}"
`;

exports[`forgetti skip > should optimize non-skipped function expression 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
import { memo as _memo } from \\"react\\";
import { $$memo as _$$memo } from \\"forgetti/runtime\\";
const _Example = _$$memo(_memo, \\"_Example\\", _values => <h1 className={_values[0]}>{_values[1]}</h1>);
const Example = function (props) {
  let _cache = _$$cache(_useMemo, 8),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = _value.className,
    _equals2 = _$$equals(_cache, 2, _value2),
    _value3 = _equals2 ? _cache[2] : _cache[2] = _value2,
    _value4 = _equals ? _cache[3] : _cache[3] = _value.children,
    _equals3 = _$$equals(_cache, 4, _value4),
    _value5 = _equals3 ? _cache[4] : _cache[4] = _value4,
    _value6 = _equals2 && _equals3 ? _cache[5] : _cache[5] = [_value3, _value5],
    _equals5 = _$$equals(_cache, 6, _value6),
    _value7 = _equals5 ? _cache[6] : _cache[6] = _value6;
  return /*@forgetti jsx*/_equals5 ? _cache[7] : _cache[7] = <_Example v={_value7} />;
};"
`;

exports[`forgetti skip > should optimize non-skipped variable declaration 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
import { memo as _memo } from \\"react\\";
import { $$memo as _$$memo } from \\"forgetti/runtime\\";
const _Example = _$$memo(_memo, \\"_Example\\", _values => <h1 className={_values[0]}>{_values[1]}</h1>);
const Example = props => {
  let _cache = _$$cache(_useMemo, 8),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = _value.className,
    _equals2 = _$$equals(_cache, 2, _value2),
    _value3 = _equals2 ? _cache[2] : _cache[2] = _value2,
    _value4 = _equals ? _cache[3] : _cache[3] = _value.children,
    _equals3 = _$$equals(_cache, 4, _value4),
    _value5 = _equals3 ? _cache[4] : _cache[4] = _value4,
    _value6 = _equals2 && _equals3 ? _cache[5] : _cache[5] = [_value3, _value5],
    _equals5 = _$$equals(_cache, 6, _value6),
    _value7 = _equals5 ? _cache[6] : _cache[6] = _value6;
  return /*@forgetti jsx*/_equals5 ? _cache[7] : _cache[7] = <_Example v={_value7} />;
};"
`;

exports[`forgetti skip > should skip skipped function declaration 1`] = `
"/* @forgetti skip */
function Example(props) {
  return <h1 className={props.className}>{props.children}</h1>;
}"
`;

exports[`forgetti skip > should skip skipped function expression 1`] = `
"const /* @forgetti skip */ExampleA = function (props) {
  return <h1 className={props.className}>{props.children}</h1>;
};
/* @forgetti skip */
const ExampleB = function (props) {
  return <h1 className={props.className}>{props.children}</h1>;
};"
`;

exports[`forgetti skip > should skip skipped variable declaration 1`] = `
"const /* @forgetti skip */ExampleA = props => {
  return <h1 className={props.className}>{props.children}</h1>;
};
/* @forgetti skip */
const ExampleB = props => {
  return <h1 className={props.className}>{props.children}</h1>;
};"
`;
